user_profile/views.py

```diff
@@ -0,0 +1,143 @@
+import datetime
+
+from django.contrib.auth.decorators import user_passes_test
+from django.contrib.auth.mixins import PermissionRequiredMixin, UserPassesTestMixin
+from django.contrib.auth.models import User
+from django.contrib import messages
+from django.http import HttpResponseRedirect, HttpResponse
+from django.shortcuts import render, get_object_or_404, redirect
+from django.urls import reverse, reverse_lazy
+from django.views.generic import DetailView, CreateView, UpdateView, DeleteView, ListView
+
+from .models import Profile, Skill, Project
+from .forms import ProfileForm, ProjectForm
+
+
+class DetailSkillView(DetailView):
+    model = Skill
+
+
+class DetailProjectView(DetailView):
+    model = Project
+
+
+class ListProjectView(ListView):
+    model = Project
+
+
+class CreateProjectView(PermissionRequiredMixin, CreateView):
+    redirect_field_name = '/'
+    permission_required = 'user_profile.add_project'
+    model = Project
+    form_class = ProjectForm
+
+
+class EditProjectView(UserPassesTestMixin, UpdateView):
+    model = Project
+    form_class = ProjectForm
+    redirect_field_name = '/'
+
+    def test_func(self):
+        return self.request.user == get_object_or_404(Project, pk=self.kwargs['pk']).manager
+
+
+class DeleteProjectView(UserPassesTestMixin, DeleteView):
+    model = Project
+    redirect_field_name = '/'
+    success_url = reverse_lazy('project_list')
+
+    # Should be same the same in DeleteProject and EditProject
+    def test_func(self):
+        return self.request.user == get_object_or_404(Project, pk=self.kwargs['pk']).manager
+
+##############################################
+
+
+def administrate_project(request, pk):
+    project = get_object_or_404(Project, pk=pk)
+    if not request.user == project.manager:
+        return HttpResponseRedirect("/")
+
+    return render(request, 'user_profile/project_admin.html', {'project': project})
+
+
+def accept_applicant(request, pk, username):
+    return manage_applicant(request, pk, username, True)
+
+
+def reject_applicant(request, pk, username):
+    return manage_applicant(request, pk, username, False)
+
+
+def apply_to_project(request, pk):
+    if request.method == 'POST':
+        user = get_object_or_404(User, username=request.user.username)
+        project = get_object_or_404(Project, pk=pk)
+
+        # if not rejected, already an applicant or already a member
```

Comment:
filter.exists might be faster
Not sure if Django queries all from database then checks for membership or actually does the membership check as a database query.

Reply:
According to what I've read on SO, get_or_404(<Object>, <key>) works like a <Object>.get(<key>) with a raise 404Error in-built if nothing is found.